barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit.hpp

```diff
@@ -124,79 +119,135 @@ template <typename FF> void Circuit<FF>::init()
         }
     }
 
-    vars[0] == FF::Const("0", this->solver);
-    vars[1] == FF::Const("1", this->solver);
+    symbolic_vars[0] == bb::fr(0);
+    symbolic_vars[1] == bb::fr(1);
+
+}
+
+template <typename FF> void Circuit<FF>::init_graph(){
+    for(uint32_t i = 0; i < this->get_num_real_vars(); i++){
+        this->graph[i] = std::vector<uint32_t>(this->get_num_real_vars(), 0);
+    }
+
+    for(uint32_t i = 0; i < this->wires_idxs.size(); i++){
+        uint32_t wl = this->wires_idxs[i][0];
+        uint32_t wr = this->wires_idxs[i][1];
+        uint32_t wo = this->wires_idxs[i][2];
+        this->graph[wl][wo] = i; // TODO(alex): sometimes it's breaking the logic. I think I might need to use multigraph structure.
+        this->graph[wr][wo] = i;
+        this->graph[wo][wl] = -i;
+        this->graph[wo][wr] = -i;
+    }
+}
+
 
-    for (auto i : public_inps) {
-        vars[i] == FF::Const(variables[i], this->solver);
+/**
+ * @brief Relaxes quadratic polynomial constraint.
+ *
+ * @param q_m multiplication selector
+ * @param q_1 l selector
+ * @param q_2 r selector
+ * @param q_3 o selector
+ * @param q_c constant
+ * @param w   witness index
+ */
+template <typename FF>
+void Circuit<FF>::quadratic_polynomial_handler(bb::fr q_m, bb::fr q_1, bb::fr q_2, bb::fr q_3, bb::fr q_c, uint32_t w){
+    bb::fr b = q_1 + q_2 + q_3;
+
+    if (q_m == 0){
+        symbolic_vars[w] == -q_c / b;
+        return;
+    }
+
+    std::pair<bool, bb::fr> d = (b*b - bb::fr(4) * q_m * q_c).sqrt();
+    if(!d.first){
+        throw std::invalid_argument("There're no roots of quadratic polynomial");
     }
+    bb::fr x1 = (-b + d.second) / (bb::fr(2) * q_m);
+    bb::fr x2 = (-b - d.second) / (bb::fr(2) * q_m);
+
+    ((Bool(symbolic_vars[w]) == Bool(FF(x1, this->solver))) |
+     (Bool(symbolic_vars[w]) == Bool(FF(x2, this->solver)))).assert_term();
 }
 
+template <typename FF> void Circuit<FF>::prepare_gates(){
+    return;
+} // TODO(alex): complete
+
 /**
  * @brief Adds all the gate constraints to the solver.
  *
  */
 template <typename FF> void Circuit<FF>::add_gates()
 {
+    this->prepare_gates();
+
     for (size_t i = 0; i < get_num_gates(); i++) {
-        FF q_m = FF::Const(selectors[i][0], this->solver);
-        FF q_1 = FF::Const(selectors[i][1], this->solver);
-        FF q_2 = FF::Const(selectors[i][2], this->solver);
-        FF q_3 = FF::Const(selectors[i][3], this->solver);
-        FF q_c = FF::Const(selectors[i][4], this->solver);
+        bb::fr q_m = selectors[i][0];
+        bb::fr q_1 = selectors[i][1];
+        bb::fr q_2 = selectors[i][2];
+        bb::fr q_3 = selectors[i][3];
+        bb::fr q_c = selectors[i][4];
 
         uint32_t w_l = wires_idxs[i][0];
         uint32_t w_r = wires_idxs[i][1];
         uint32_t w_o = wires_idxs[i][2];
 
-        // Binary gate. Relaxes the solver.
+        // Binary gate. Relaxes the solver. 
         // TODO(alex): Probably we can add other basic gates here too to relax the stuff.
```

Comment:
You could add the basic xor and add gates now that you know what they are (the ones that connect individual bits)